dgl_ptm/dgl_ptm/model/initialize_model.py

```diff
@@ -315,21 +361,75 @@ def _initialize_agents_tec(self):
         for i in range(len(self.technology_levels)):
             technology_mask = agentsTecLevel == i
             agentsGamma[technology_mask] = self.gamma_vals[i]
-            agentsCost[technology_mask] = self.cost_vals[i]   
+            agentsCost[technology_mask] = self.cost_vals[i]
         return agentsTecLevel, agentsGamma, agentsCost
 
     def step(self):
         try:
             self.step_count +=1
-            ptm_step(self.model_graph,self.model_data,self.step_count,self.steering_parameters)
-        except:
-            #TODO add model dump here. Also check against previous save to avoid overwriting
+            print(f'performing step {self.step_count} of {self.step_target}')
+            ptm_step(self.model_graph, self.model_data, self.step_count, self.steering_parameters)
 
+            # store each step
+            self.inputs[self.step_count] = {
+                'model_graph': copy.deepcopy(self.model_graph),
+                'model_data': self.model_data,
+                'generator_state': generator.get_state(),
+            }
 
+        except:
+            #TODO add model dump here. Also check against previous save to avoid overwriting
             raise RuntimeError(f'execution of step failed for step {self.step_count}')
 
-
     def run(self):
+        """ run the model for each step until the step_target is reached."""
+
+        self.model_graph = copy.deepcopy(self.inputs[self.step_count]['model_graph'])
+        self.model_data = self.inputs[self.step_count]['model_data']
+        generator.set_state(self.inputs[self.step_count]['generator_state'])
+
         while self.step_count < self.step_target:
-            print(f'performing step {self.step_count+1} of {self.step_target}')
             self.step()
+
+        if self.savestate:
+            path = f'./{self._model_identifier}'
+            _save_model(path, self.inputs)
+
+
+def _save_model(path, dictionary):
+    """ save the model_graph and generator_state in files."""
```

Comment:
Is this correct that we are creating a dictionary saving the all the model states for each step, and then writing to file/updatiing the ffile version on demand? This seems problematic in terms of memory implications.

Reply:
Just checked. The current implementation indeeds creates a dictionary of dictionary wit all saved sates. While it may not be a problem at the number of agents and steps used in testing this will become an issue in production and should be addressed

Reply:
It is fixed now. Now it only saves every X step(overwriting the previous state) and loads the last saved step. So the memory and disk space would not be a problem.